compiler/ir/backend.common/src/org/jetbrains/kotlin/backend/common/lower/FlattenStringConcatenationLowering.kt

```diff
@@ -73,11 +75,13 @@ class FlattenStringConcatenationLowering(val context: CommonBackendContext) : Fi
             return when (expression) {
                 is IrStringConcatenation -> true
                 is IrCall -> {
-                    val dispatchReceiver = expression.dispatchReceiver
-                    dispatchReceiver != null &&
-                            dispatchReceiver.type.isString() &&
-                            expression.symbol.owner.name == PLUS_NAME &&
-                            expression.type.isString() &&
+                    val function = expression.symbol.owner
+                    val receiver = expression.dispatchReceiver ?: expression.extensionReceiver
+                    receiver != null &&
+                            receiver.type.isStringClassType() &&
+                            function.getPackageFragment()?.fqName == KotlinBuiltIns.BUILT_INS_PACKAGE_FQ_NAME &&
```

Comment:
@sfs Please use `parent` directly and check that it's a package fragment instead of using `getPackageFragment` which walks up until it finds a package fragment (which is not precise enough and _could_ result in incorrect behavior here for a member function, but it seems impossible due to other conditions).

Reply:
@udalov That was actually on purpose, since there are two functions we check for, `kotlin.String.plus` for non-nullable strings and `kotlin.plus` for nullable strings. If you prefer I could instead check for these two cases explicitly.

Reply:
Thanks for the clarification. Yes, I'd prefer that since it's more explicit and avoids confusion with other functions named `plus` in the standard library (e.g. if `receiver`'s initializer is refactored to `expression.extensionReceiver ?: expression.dispatchReceiver`, we could incorrectly identify some member extension on `String` named `plus` in an irrelevant class here)